src/dense.cc

```diff
@@ -0,0 +1,295 @@
+#include "dense.hh"
+
+#include <iostream>
+#include <stdio.h>
+#include <stdexcept>
+#include <string>
+#include <omp.h>
+
+#include <math.h>
+#include <typeinfo>
+
+#if !defined(R123_NO_SINCOS) && defined(__APPLE__)
+/* MacOS X 10.10.5 (2015) doesn't have sincosf */
+// use "-D __APPLE__" as a compiler flag to make sure this is hit.
+#define R123_NO_SINCOS 1
+#endif
+
+#if R123_NO_SINCOS /* enable this if sincos and sincosf are not in the math library */
+static inline void sincosf(float x, float *s, float *c) {
+    *s = sinf(x);
+    *c = cosf(x);
+}
+
+static inline void sincos(double x, double *s, double *c) {
+    *s = sin(x);
+    *c = cos(x);
+}
+#endif /* sincos is not in the math library */
+
+
+// The following two functions are part of NVIDIA device side math library.
+// Random123 relies on them in both host and device sources.  We can work
+// around this by defining them when not compiling device code.
+#if !defined(__CUDACC__)
+
+static inline void sincospif(float x, float *s, float *c) {
+    const float PIf = 3.1415926535897932f;
+    sincosf(PIf*x, s, c);
+}
+
+static inline void sincospi(double x, double *s, double *c) {
+    const double PI = 3.1415926535897932;
+    sincos(PI*x, s, c);
+}
```

Comment:
This is to handle compilation issues that can arise when compiling on macOS and/or compiling with CUDA.